File: templates/cookbook/route-handler.py
"""
route-handler.py — CopilotForge Cookbook Recipe

WHAT THIS DOES:
    Demonstrates a FastAPI route handler with Pydantic input validation,
    dependency injection, proper HTTP status codes, and structured error
    responses.

WHEN TO USE THIS:
    When building FastAPI applications and you want consistent request
    validation, error responses, and typed route handlers.

HOW TO RUN:
    1. pip install fastapi uvicorn
    2. uvicorn cookbook.route-handler:app --reload

PREREQUISITES:
    - Python 3.10+
    - fastapi >= 0.100
    - uvicorn >= 0.23
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Annotated

import uvicorn
from fastapi import Depends, FastAPI, HTTPException, Query, status
from pydantic import BaseModel, Field

# --- Schemas ---


class Create{{resource_type}}Request(BaseModel):
    name: str = Field(min_length=1, max_length=100, description="Name")
    description: str | None = Field(default=None, max_length=500, description="Optional description")
    price: float = Field(gt=0, description="Price must be positive")
    tags: list[str] = Field(default_factory=list, max_length=10, description="Up to 10 tags")


class Update{{resource_type}}Request(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    price: float | None = Field(default=None, gt=0)
    tags: list[str] | None = Field(default=None, max_length=10)


class {{resource_type}}Response(BaseModel):
    id: str
    name: str
    description: str | None = None
    price: float
    tags: list[str]
    created_at: str


class PaginationInfo(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int


class {{resource_type}}ListResponse(BaseModel):
    data: list[{{resource_type}}Response]
    pagination: PaginationInfo


# --- In-Memory Store (replace with real database) ---

# TODO: Replace with actual database queries against {{db_connection_string}}.
store: dict[str, {{resource_type}}Response] = {}

# --- Dependencies ---


class ListParams:
    """Dependency for pagination and search query params."""

    def __init__(
        self,
        page: Annotated[int, Query(ge=1, description="Page number")] = 1,
        page_size: Annotated[int, Query(ge=1, le=100, alias="pageSize")] = 20,
        search: Annotated[str | None, Query(description="Search term")] = None,
    ) -> None:
        self.page = page
        self.page_size = page_size
        self.search = search


# --- Helper ---


def get_or_404(item_id: str) -> {{resource_type}}Response:
    item = store.get(item_id)
    if item is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NOT_FOUND", "message": f"{{resource_type}} '{item_id}' not found"},
        )
    return item


# --- App & Routes ---

app = FastAPI(title="{{project_name}} API")


@app.get("/api/{{resource_path}}", response_model={{resource_type}}ListResponse)
async def list_items(params: ListParams = Depends()) -> {{resource_type}}ListResponse:
    results = list(store.values())

    if params.search:
        term = params.search.lower()
        results = [
            item
            for item in results
            if term in item.name.lower() or (item.description and term in item.description.lower())
        ]

    total = len(results)
    start = (params.page - 1) * params.page_size
    paged = results[start : start + params.page_size]

    return {{resource_type}}ListResponse(
        data=paged,
        pagination=PaginationInfo(
            page=params.page,
            page_size=params.page_size,
            total=total,
            total_pages=max(1, -(-total // params.page_size)),
        ),
    )


@app.get("/api/{{resource_path}}/{item_id}", response_model=dict[str, {{resource_type}}Response])
async def get_item(item_id: str) -> dict[str, {{resource_type}}Response]:
    return {"data": get_or_404(item_id)}


@app.post("/api/{{resource_path}}", response_model=dict[str, {{resource_type}}Response], status_code=status.HTTP_201_CREATED)
async def create_item(body: Create{{resource_type}}Request) -> dict[str, {{resource_type}}Response]:
    item_id = str(uuid.uuid4())

    item = {{resource_type}}Response(
        id=item_id,
        name=body.name,
        description=body.description,
        price=body.price,
        tags=body.tags,
        created_at=datetime.now(timezone.utc).isoformat(),
    )

    store[item_id] = item
    return {"data": item}


@app.patch("/api/{{resource_path}}/{item_id}", response_model=dict[str, {{resource_type}}Response])
async def update_item(item_id: str, body: Update{{resource_type}}Request) -> dict[str, {{resource_type}}Response]:
    item = get_or_404(item_id)
    update_data = body.model_dump(exclude_unset=True)
    updated = item.model_copy(update=update_data)
    store[item_id] = updated
    return {"data": updated}


@app.delete("/api/{{resource_path}}/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_item(item_id: str) -> None:
    get_or_404(item_id)
    del store[item_id]


# --- Startup ---


def main() -> None:
    # TODO: Replace host/port with your preferred values.
    uvicorn.run(app, host="0.0.0.0", port={{server_port}})


if __name__ == "__main__":
    main()

File: templates/cookbook/route-handler.ts
/**
 * route-handler.ts — CopilotForge Cookbook Recipe
 *
 * WHAT THIS DOES:
 *   Demonstrates an Express route handler with Zod input validation,
 *   typed request/response, centralized error middleware, and proper
 *   HTTP status codes.
 *
 * WHEN TO USE THIS:
 *   When building Express APIs and you want consistent request validation,
 *   error responses, and typed handlers.
 *
 * HOW TO RUN:
 *   1. npm install express zod
 *   2. npm install -D @types/express
 *   3. npx ts-node cookbook/route-handler.ts
 *
 * PREREQUISITES:
 *   - Node.js 18+
 *   - TypeScript 5+
 *   - express ^4.18
 *   - zod ^3.22
 */

import express, { Request, Response, NextFunction, Router } from "express";
import { z, ZodError, ZodSchema } from "zod";

// --- Validation Middleware ---

function validate(schema: ZodSchema, source: "body" | "query" | "params" = "body") {
  return (req: Request, res: Response, next: NextFunction): void => {
    const result = schema.safeParse(req[source]);

    if (!result.success) {
      const errors = result.error.errors.map((e) => ({
        field: e.path.join("."),
        message: e.message,
      }));

      res.status(400).json({
        error: {
          code: "VALIDATION_ERROR",
          message: "Request validation failed",
          details: errors,
        },
      });
      return;
    }

    req[source] = result.data;
    next();
  };
}

// --- Error Middleware ---

class RouteError extends Error {
  readonly statusCode: number;
  readonly code: string;

  constructor(message: string, statusCode: number, code: string) {
    super(message);
    this.name = "RouteError";
    this.statusCode = statusCode;
    this.code = code;
  }
}

function errorHandler(err: Error, _req: Request, res: Response, _next: NextFunction): void {
  if (err instanceof RouteError) {
    res.status(err.statusCode).json({
      error: { code: err.code, message: err.message },
    });
    return;
  }

  if (err instanceof ZodError) {
    const errors = err.errors.map((e) => ({
      field: e.path.join("."),
      message: e.message,
    }));
    res.status(400).json({
      error: { code: "VALIDATION_ERROR", message: "Validation failed", details: errors },
    });
    return;
  }

  console.error("[ERROR]", err);
  res.status(500).json({
    error: { code: "INTERNAL_ERROR", message: "An unexpected error occurred" },
  });
}

// --- Schemas ---

const create{{resource_type}}Schema = z.object({
  name: z.string().min(1, "Name is required").max(100, "Name must be 100 chars or less"),
  description: z.string().max(500).optional(),
  price: z.number().positive("Price must be positive"),
  tags: z.array(z.string()).max(10).default([]),
});

const update{{resource_type}}Schema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().max(500).optional(),
  price: z.number().positive().optional(),
  tags: z.array(z.string()).max(10).optional(),
});

const listQuerySchema = z.object({
  page: z.coerce.number().int().min(1).default(1),
  pageSize: z.coerce.number().int().min(1).max(100).default(20),
  search: z.string().optional(),
});

const idParamSchema = z.object({
  id: z.string().uuid("Invalid ID format"),
});

// --- Types ---

type Create{{resource_type}}Input = z.infer<typeof create{{resource_type}}Schema>;
type Update{{resource_type}}Input = z.infer<typeof update{{resource_type}}Schema>;
type ListQuery = z.infer<typeof listQuerySchema>;

interface {{resource_type}} {
  id: string;
  name: string;
  description?: string;
  price: number;
  tags: string[];
  createdAt: string;
}

// --- In-Memory Store (replace with real database) ---

// TODO: Replace with actual database queries against {{db_connection_string}}.
const store: Map<string, {{resource_type}}> = new Map();

// --- Route Handlers ---

const router = Router();

router.get("/{{resource_path}}", validate(listQuerySchema, "query"), (req: Request, res: Response) => {
  const query = req.query as unknown as ListQuery;

  let results = Array.from(store.values());

  if (query.search) {
    const term = query.search.toLowerCase();
    results = results.filter(
      (item) =>
        item.name.toLowerCase().includes(term) ||
        item.description?.toLowerCase().includes(term)
    );
  }

  const total = results.length;
  const start = (query.page - 1) * query.pageSize;
  const paged = results.slice(start, start + query.pageSize);

  res.json({
    data: paged,
    pagination: {
      page: query.page,
      pageSize: query.pageSize,
      total,
      totalPages: Math.ceil(total / query.pageSize),
    },
  });
});

router.get("/{{resource_path}}/:id", validate(idParamSchema, "params"), (req: Request, res: Response, next: NextFunction) => {
  const { id } = req.params;
  const item = store.get(id);

  if (!item) {
    return next(new RouteError(`{{resource_type}} '${id}' not found`, 404, "NOT_FOUND"));
  }

  res.json({ data: item });
});

router.post("/{{resource_path}}", validate(create{{resource_type}}Schema, "body"), (req: Request, res: Response) => {
  const input = req.body as Create{{resource_type}}Input;
  const id = crypto.randomUUID();

  const item: {{resource_type}} = {
    id,
    name: input.name,
    description: input.description,
    price: input.price,
    tags: input.tags,
    createdAt: new Date().toISOString(),
  };

  store.set(id, item);
  res.status(201).json({ data: item });
});

router.patch(
  "/{{resource_path}}/:id",
  validate(idParamSchema, "params"),
  validate(update{{resource_type}}Schema, "body"),
  (req: Request, res: Response, next: NextFunction) => {
    const { id } = req.params;
    const item = store.get(id);

    if (!item) {
      return next(new RouteError(`{{resource_type}} '${id}' not found`, 404, "NOT_FOUND"));
    }

    const input = req.body as Update{{resource_type}}Input;

    if (input.name !== undefined) item.name = input.name;
    if (input.description !== undefined) item.description = input.description;
    if (input.price !== undefined) item.price = input.price;
    if (input.tags !== undefined) item.tags = input.tags;

    res.json({ data: item });
  }
);

router.delete("/{{resource_path}}/:id", validate(idParamSchema, "params"), (req: Request, res: Response, next: NextFunction) => {
  const { id } = req.params;

  if (!store.has(id)) {
    return next(new RouteError(`{{resource_type}} '${id}' not found`, 404, "NOT_FOUND"));
  }

  store.delete(id);
  res.status(204).send();
});

// --- App Setup ---

async function main() {
  const app = express();
  app.use(express.json());

  app.use("/api", router);
  app.use(errorHandler);

  // TODO: Replace port with your preferred port.
  const PORT = parseInt(process.env.PORT ?? "{{server_port}}", 10);
  app.listen(PORT, () => {
    console.log(`[{{project_name}}] Server listening on http://localhost:${PORT}`);
  });
}

main().catch(console.error);
